src/installer/corehost/cli/hostpolicy/hostpolicy.cpp

```diff
@@ -744,6 +744,7 @@ SHARED_API int HOSTPOLICY_CALLTYPE corehost_initialize(const corehost_initialize
             rc = StatusCode::Success_DifferentRuntimeProperties;
     }
 
+    auto version_lo = context_contract->version;
```

Comment:
I think this is a problem. Shipped versions of `hostfxr` (3.0) would not initialize this value on input. So `hostpolicy` must not rely on the value being correct (it's effectively random data).

We will need to find a different way for hostfxr to pass along the version information. The only one I can think of is to add a new bit to the `initialization_options_t` - something like `versioned_contract` or something similar which would tell hostpolicy that it's OK to look at the version field - in fact in that case I think `hostfxr` should initialize all the fields, so that in the future we can use them to communicate intent both ways if needed.

This is unfortunate - we should have set the version in 3.0 already, the contract as is, is effectively broken because the allocated size comes from `hostfxr`, but it's filled with `hostpolicy` without telling it how big the buffer is. I'm starting to question if we should introduce a new API just to fix this.

Reply:
I initially considered passing the size of the structure into `initialization_options_t` but I thought this would be easier. I'm not used to working in code that doesn't *always* initialize variables, so I forgot you couldn't rely on it being all zeros. I'm going to hold off on making any more changes until you guys decide how you want to handle this.

Reply:
@elinor-fung @AaronRobinsonMSFT @swaroop-sridhar any ideas?

The short story: in 3.0 hostfxr calls hostpolicy's initialize with a pointer to a structure which hostpolicy fills in. The structure has a version field, but it's only initialized on the hostpolicy side. So if we decide to add more fields to the struct in 5.0, there's no way to pass the size of the allocated struct from hostfxr to hostpolicy - since hostpolicy can't trust the version field on input (as 3.0 hostfxr didn't initialize it).

Options:
* Introduce a new initialize API on hostpolicy which should probably have one more parameter - the size of the struct. Just like any other out buffer param should have.
* Add a new bit to the initialize options which would tell hostpolicy that newer hostfxr is calling and that the version field in the struct is initialized and thus hostpolicy can trust it (basically a version flag).
* ???

We could also do this somehow via the global initialization of hostpolicy (load), but I don't like that very much.

Reply:
I am unclear why this is an issue. That context is intentionally an out parameter that informs hostfxr of the contract that hostpolicy has. This API isn't a conversation it is a question - "tell me what you support". The caller needs to check the version and can then use the struct however it is filled out.

Reply:
The problem is the size of the out parameter. The memory is caller allocated (the caller passes a pointer to a piece of memory) but the method doesn't take the size of the allocated memory. Right now it works because both caller and callee agree on the structure. If we want to version that structure (it has a version field, so it was designed for this), we also need to somehow pass how much memory is available, so that the callee can write more data into it, if there's space. Currently we don't do that, so the callee must always write exactly the same number of bytes it always did.

Reply:
Spoke offline and finally understand the issue. I would suggest providing a new flag that is something like `version_set` or something like that.

Reply:
Offline discussion with @AaronRobinsonMSFT:
* It is possible to have older `hostfxr` talking to a newer `hostpolicy` (private install of older runtime - running dotnet.exe from it - DOTNET_MULTILEVEL_LOOKUP=1 which is the default - trying to run an app - the older hostfxr will now also search global install location for frameworks - if the app allows it it may roll-forward to a newer framework from global install - that has hostpolicy which is newer than the hostfxr).

* Currently hostpolicy doesn't know what version of hostfxr it's talking to (we have version fields on the host_interface_t but they're effectively useless for this purpose, at least the way they're used now).

* Would like: Add a full version of hostfxr into the host_interface_t - could be used by hostpolicy in case we need to quirk something (let's hope not, but seeing this we might). Pass it along to the runtime which could trace it for auditing purposes, or diag purposes and so on (currently there's no easy way to tell the version of hostfxr used to run a given app). This should be done in a separate PR as it's not needed for this change.

* Solve this current problem by adding a bit to the `options` parameter for `corehost_initialize` - something like `initialization_options_t::context_contract_version_set = 0x80000000`. hostpolicy will check that flag and if set it will assume the `corehost_context_contract::version` field is initialized on input (what this PR already does) and will use it to determine if it can fill the struct with the new fields or not. If the bit is not set it will not read from the buffer at all (random data there) and instead assume the "old" version of the structure - basically it can assume `version=sizeof(size_t)*7`.

Reply:
I can start trying to add that bit to `intialization_options_t` (it's misspelled :( ). (In my original response I didn't look at the code and assumed that this was the global initialization struct, my initial thought was to add something to `host_interface_t` but this should work just as well).

Reply:
>intialization_options_t (it's misspelled :( )

You can fix that spelling if you like. It doesn't represent breaking compat change.

Reply:
Can you please also fix the spelling - that is perfectly doable as it's only a build-time thing (the hostpolicy header is internal, nobody but hostfxr should be calling hostpolicy directly).